Observer/CancelOrderItemObserver.php

```diff
@@ -150,31 +160,28 @@ public function execute(EventObserver $observer): void
         $this->placeReservationsForSalesEvent->execute($itemsToCancel, $salesChannel, $salesEvent);
 
         $order = $orderItem->getOrder();
-
-        // Source selection happens by default on shipment, so only shipments contain information about the
-        // inventory source.
-        // @TODO add support for multi source stock replenishment on cancellation
-        if ($this->isSingleSourceMode->execute()) {
-            $sourceCode = $this->defaultSourceProvider->getCode();
-
-            $itemsToDeduct = [];
-            foreach ($itemsToCancel as $itemToCancel) {
-                $itemsToDeduct[] = $this->itemToDeductFactory->create([
-                    'sku' => $itemToCancel->getSku(),
-                    'qty' => -$itemToCancel->getQuantity(),
-                ]);
+        $sourceSelectionItems = $this->getSourceSelectionResultFromOrder->execute($order)->getSourceSelectionItems();
+
+        foreach ($itemsToCancel as $itemToCancel) {
+            foreach ($sourceSelectionItems as $sourceSelectionItem) {
+                if ($sourceSelectionItem->getSku() === $itemToCancel->getSku()) {
```

Comment:
Once the above is addressed you shouldn't need this nested loop. You could even add functionality to your repo to fetch a source item by sku?